nmoscommon/auth/auth_registrar.py

```diff
@@ -0,0 +1,200 @@
+            #     data, token_href, client_id, client_secret
+            # ))  # Only for Testing
+
+            token_req = requests.post(token_href,
+                                      data=data,
+                                      auth=(client_id, client_secret),
+                                      timeout=0.5,
+                                      proxies={'http': ''})
+            token_req.raise_for_status()
+            print("Token Request Successful")
+            return token_req.json()
+        except Exception as e:
+            print("Bearer Token Request Failed. {}".format(e))
+            return False
+
+    def token_request_password(self, username, password, scope):
+        return self.token_request(grant_type="password", username=username, password=password, scope=scope)
+
+    def token_request_code(self, code, scope):
+        return self.token_request(grant_type="authorization_code", auth_code=code, scope=scope)
+
+    def token_request(self, grant_type, username=None, password=None, auth_code=None, scope=None):
+        data = None
+
+        if grant_type == "password" and username is not None and password is not None:
+            data = {
+                    "grant_type": grant_type,
+                    "username": username,
+                    "password": password
+            }
+        elif grant_type == "authorization_code" and auth_code is not None:
+            if self.auth_reg is not None:
+                data = {
+                    "grant_type": grant_type,
+                    "code": auth_code,
+                    "redirect_uri": self.auth_reg.redirect_uri,
+                    "client_id": self.auth_reg.client_id
+                }
+            else:
+                client_id, client_secret = get_credentials_from_file(CREDENTIALS_PATH)
+                data = {
+                    "grant_type": grant_type,
+                    "code": auth_code,
+                    "redirect_uri": None,
+                    "client_id": client_id
+                }
+        if data is None:
+            raise Exception("Invalid Credentials for the supplied Grant Type")
+
+        if scope is not None:
+            data.update({'scope': scope})
+
+        return self._auth_request(data)
+
+
+if __name__ == "__main__":  # pragma: no cover
+    auth_reg = AuthRegistrar(
+        client_name="test_oauth_client",
+        client_uri="www.example.com",
+        allowed_scope="is04",
+        redirect_uri="www.app.example.com",
+        allowed_grant="password\nauthorization_code",  # Authlib only seems to accept grants seperated with newline chars
+        allowed_response="code",
+        auth_method="client_secret_basic",
+        certificate=None
+    )
+    if auth_reg.initialised is True:
+        auth_req = AuthRequestor(auth_reg)
+        token = auth_req.token_request_password(username='dannym', password='password', scope='is04')
```

Comment:
Again - we need to figure out how this should actually work...

Reply:
switching to the credentials grant would eliminate the need for username / password?